api/pin.go

```diff
@@ -0,0 +1,423 @@
+		err := p.db.Set(context.TODO(), chunk.ModePinChunk, chunkAddr)
+		if err != nil {
+			log.Error("Could not pin chunk. Address " + hex.EncodeToString(chunkAddr))
+			return err
+		} else {
+			log.Trace("Pinning chunk", "Address", hex.EncodeToString(chunkAddr))
+		}
+		return nil
+	}
+	p.WalkChunksFromRootHash(rootHash, isRaw, credentials, walkerFunction)
+
+	// Check if the root hash is already pinned
+	isFilePinned := p.db.IsFilePinned(addr)
+	if !isFilePinned {
+		if isRaw {
+			err = p.db.Set(context.TODO(), chunk.ModeStoreRootHashForRawFile, addr)
+		} else {
+			err = p.db.Set(context.TODO(), chunk.ModeStoreRootHashForNormalFile, addr)
+		}
+		if err != nil {
+			// TODO: if this happens, we should go back and revert the entire file's chunks
+			log.Error("Could not unpin root chunk. Address " + fmt.Sprintf("%0x", addr))
+			return err
+		}
+	}
+	return nil
+}
+
+// UnPinFiles is used to unpin an already pinned file. It takes the root
+// hash of the file and walks down the merkle tree unpinning all the chunks
+// that are encountered on the way. The pre-requisite is that the file should
+// have been already pinned using the PinFiles function. This function can
+// be called only from an external command.
+func (p *PinApi) UnpinFiles(rootHash string, credentials string) {
+
+	addr, err := hex.DecodeString(rootHash)
+	if err != nil {
+		log.Error("Error decoding root hash" + err.Error())
+		return
+	}
+
+	isRawInDB, err := p.db.IsPinnedFileRaw(addr)
+	if err != nil {
+		log.Error("Root hash is not pinned" + err.Error())
+		return
+	}
+
+	// Walk the root hash and unpin all the chunks
+	walkerFunction := func(ref storage.Reference) error {
+		chunkAddr := p.removeDecryptionKeyFromChunkHash(ref)
+		err := p.db.Set(context.TODO(), chunk.ModeUnPinChunk, chunkAddr)
+		if err != nil {
+			log.Error("Could not unpin chunk. Address " + hex.EncodeToString(chunkAddr))
+			return err
+		} else {
+			log.Trace("Unpinning chunk", "Address", fmt.Sprintf("%0x", chunkAddr))
+		}
+		return nil
+	}
+	p.WalkChunksFromRootHash(rootHash, isRawInDB, credentials, walkerFunction)
+
+	// Check if the root chunk exists in pinIndex
+	// If it is not.. then the pin counter became 0
+	// so remove the root hash from pinFilesIndex
+	isRootChunkPinned := p.db.IsChunkPinned(p.removeDecryptionKeyFromChunkHash(addr))
+	if !isRootChunkPinned {
+		err := p.db.Set(context.TODO(), chunk.ModeRemoveRootHash, addr)
+		if err != nil {
+			// TODO: if this happens, we should go back and revert the entire file's chunks
+			log.Error("Could not unpin root chunk. Address " + fmt.Sprintf("%0x", addr))
+		}
+	}
+}
+
+// ListPinFiles functions logs information of all the files that are pinned
+// in the current local node. It displays the root hash of the pinned file
+// or collection. It also display two vital information
+//     1) Size of the pinned file or collection
+//     2) the number of times that particular file or collection is pinned.
+func (p *PinApi) ListPinFiles() {
```

Comment:
I don't understand how it can used to attack a remote swarm node.
The only issue in this function is the _GetPinCounterOfChunk_ function which will cause CPU drain if the no of chunks in the file is  large. I could not avoid this because we don't have an efficient way to store these information in index. The Set function takes only address as argument and no other information is possible to be sent apriori.  We need to seriously start thinking  about localstore as a generic store.

Also this information is important as it  not only lists all the files but say how many chunks are part of this file. This is useful if you are running a pinning server as it will give you an idea about the pinned files disk utilization.

Reply:
Yes, you can cause CPU drain on iteration over a large set of files.

And for set, you are right, maybe there is a need to set additional data to the chunk, as Put is storing the whole chunk. Up until now, Set had to mark only the state of synced chunk. But maybe we can extend it with an argument that accepts values like for pinning.

Reply:
ok. So i am leaving it like this now with a TODO saying we need to get rid if this iteration.

Reply:
Cool, that is great. Thank you.